07-MemoryManagement/module/mymodule.c

```diff
@@ -0,0 +1,184 @@
+#include <linux/init.h>
+#include <linux/module.h>
+#include <linux/device.h>
+#include <linux/err.h>
+#include <linux/uaccess.h>
+#include <linux/timer.h>
+#include <linux/jiffies.h>
+#include <linux/sched.h>
+#include <linux/time.h>
+#include <linux/fs.h>
+#include <linux/string.h>
+#include <linux/slab.h>
+#include <linux/kernel.h>
+
+#define MYDEV_NAME	"MemoryManagement"
+#define SYSDIR_NAME	"my_memory_management"
+#define KMALLOC_FILE_NAME	"kmalloc-test"
+#define KZALLOC_FILE_NAME	"kzalloc-test"
+#define VMALLOC_FILE_NAME	"vmalloc-test"
+#define GETFREEPAGES_FILE_NAME	"get_free_pages-test"
+#define SEC_TO_NS(x) (x * (time_t)1E9L)
+
+static struct device pdev = { .init_name = MYDEV_NAME };
+static struct kobject *mykobj;
+
+static u64 my_pow(u64 base, unsigned int exp)
+{
+	u64 result = 1;
+
+	while (exp) {
+		if (exp & 1)
+			result *= base;
+		exp >>= 1;
+		base *= base;
+	}
+
+	return result;
+}
+
+static ssize_t show_mem(struct kobject *kobj, struct attribute *attr, char *buf)
+{
+	int ret = 0;
+	u64 allocsize = 0, relalloc = 0, relfree = 0;
+	int i = 0, j = 0;
+	void *p = NULL;
+	u64 j1 = 0, j2 = 0;
+	struct timespec ts = {};
+
+	ret = sprintf(buf, "Type mem: %s\n", attr->name);
+	ret += sprintf(buf+ret,
+		"2^X\tBuffer size(bytes)\tAllocation time(nS)\tFreeing time(nS)\n");
+
+	for (i = 0, j = 0; i < 64; j++) {
+		if (j == 2) {
+			i++;
+			j = 0;
+		}
+
+		if (strcmp(attr->name, GETFREEPAGES_FILE_NAME) == 0)
+			allocsize = i;
+		else
+			allocsize = my_pow(2, i) + j;
```

Comment:
What is the reason for using pow function for base 2? It can be implemented using shifting to left.